module/zfs/dmu_tx.c

```diff
@@ -154,16 +147,37 @@ dmu_tx_hold_object_impl(dmu_tx_t *tx, objset_t *os, uint64_t object,
 	return (txh);
 }
 
+static dmu_tx_hold_t *
+dmu_tx_hold_object_impl(dmu_tx_t *tx, objset_t *os, uint64_t object,
+    enum dmu_tx_hold_type type, uint64_t arg1, uint64_t arg2)
+{
+	dnode_t *dn = NULL;
+	dmu_tx_hold_t *txh;
+	int err;
+
+	if (object != DMU_NEW_OBJECT) {
+		err = dnode_hold(os, object, tx, &dn);
+		if (err) {
+			tx->tx_err = err;
+			return (NULL);
+		}
+	}
+	txh = dmu_tx_hold_dnode_impl(tx, os, dn, type, arg1, arg2);
+	if (txh && txh->txh_dnode)
```

Comment:
On illumos, we try to explicitly convert things to boolean, e.g. `txh != NULL && txh->txh_dnode != NULL` here.  Not sure if you do that for ZoL but I find it cleaner.